sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+const char* SKEL_DIR = "/etc/skel";
+const char* USER_HOME = "/home/%1";
+const int HOME_MODE = 0700;
+const int QUIT_TIMEOUT = 60 * 1000; // One minute quit timeout
+
+SailfishUserManager::SailfishUserManager(QObject *parent) : QObject(parent)
+{
+    qDBusRegisterMetaType<SailfishUser>();
+    qDBusRegisterMetaType<QList<SailfishUser>>();
+
+    QDBusConnection connection = QDBusConnection::systemBus();
+    if (!connection.registerObject(DBUS_OBJECT_PATH, this)) {
+        qFatal("Cannot register D-Bus object at %s", DBUS_OBJECT_PATH);
+    }
+
+    if (!connection.registerService(DBUS_INTERFACE)) {
+        qFatal("Cannot register D-Bus service at %s", DBUS_INTERFACE);
+    }
+
+    new UsermanagerAdaptor(this);
+
+    m_timer = new QTimer(this);
+    connect(m_timer, SIGNAL(timeout()), qApp, SLOT(quit()));
+    m_timer->start(QUIT_TIMEOUT);
+}
+
+QList<SailfishUser> SailfishUserManager::users(const QDBusMessage& message)
+{
+    m_timer->start();
+    QList<SailfishUser> rv;
+
+    struct group *grent = getgrnam(USER_GROUP);
+    if (grent) {
+        for (int i = 0; grent->gr_mem[i]; i++) {
+            SailfishUser user;
+            user.user = grent->gr_mem[i];
+
+            struct passwd *pw = getpwnam(grent->gr_mem[i]);
+            if (pw && pw->pw_gecos)
+                user.name = pw->pw_gecos;
+
+            rv.append(user);
+        }
+    } else {
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Getting user group failed"));
+    }
+
+    return rv;
+}
+
+bool SailfishUserManager::addGroup(const QString &group)
+{
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    bool rv = true;
+
+    if (!context) {
+        qWarning() << "Error creating context:" << lu_strerror(error);
+        return false;
+    }
+
+    struct lu_ent *ent_group = lu_ent_new();
+    if (lu_group_default(context, group.toLocal8Bit(), false, ent_group)) {
+        if (!lu_group_add(context, ent_group, &error)) {
+            qWarning() << "Adding group failed:" << lu_strerror(error);
+            rv = false;
+        }
+    } else {
+        qWarning() << "Getting group defaults failed";
+        rv = false;
+    }
+
+    lu_ent_free(ent_group);
+    lu_end(context);
+
+    return rv;
+}
+
+bool SailfishUserManager::addUserToGroup(const QString& user, const QString& group)
+{
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    if (!context) {
+        qWarning() << "Error creating context";
+        return false;
+    }
+
+    bool rv = true;
+    struct lu_ent *ent = lu_ent_new();
+    if (lu_group_lookup_name(context, group.toLocal8Bit(), ent, &error)) {
+        GValue *value = g_slice_new0(GValue);
+        g_value_init(value, G_TYPE_STRING);
+        g_value_set_string(value, user.toLocal8Bit());
```

Comment:
These toLocal8Bit() things depend on having locale correctly set up. Would it be better / worse to use toUtf8() instead?